File: pkg/swag/bank_account.go
package swag

import (
	"github.com/monetr/monetr/pkg/models"
	"time"
)

type AlwaysBankAccount struct {
	// The balance available in the account represented as whole cents. This is typically the current balance minus the
	// total value of all pending transactions. This value is not calculated in the API and is retrieved from Plaid or
	// maintained manually for manual links.
	AvailableBalance int64 `json:"availableBalance" example:"102356"`
	// The current balance in the account as whole cents without taking into consideration any pending transactions.
	CurrentBalance int64 `json:"currentBalance" example:"102400"`
	// Last 4 digits of the bank account's account number. We do not store the full bank account number or any other
	// sensitive account information.
	Mask string `json:"mask" pg:"mask" example:"9876"`
	// Name of the account, this is different than the `originalName`. This field can be changed later on while the
	// `originalName` field cannot be changed once the account is created.
	Name string `json:"name,omitempty" example:"Checking Account"`
}

type UpdateBankAccountRequest struct {
	AlwaysBankAccount
}

type CreateBankAccountRequest struct {
	AlwaysBankAccount
	// The numeric Id of the Link this bank account is associated with, if the link is manual then bank bank accounts
	// can be created manually via the API. If the Link is associated with Plaid though then bank accounts can only be
	// created through the Plaid interface. At the time of writing this there is not a way to add or remove a bank
	// account from an existing Plaid Link.
	LinkId uint64 `json:"linkId" example:"2345" validate:"required"`
	// Account Type can be; depository, credit, loan, investment or other. At the time of writing this the application
	// will only support depository. Other types may be supported in the future.
	Type models.BankAccountType `json:"accountType" example:"depository" validate:"required"`
	// Sub Type can have numerous values, but given that the application currently only supports depository the most
	// common values you will see or use are; checking and savings. Other supported types (albeit untested) are; hsa,
	// cd, money market, paypal, prepaid, cash management and ebt.
	// More information on these can be found here: https://plaid.com/docs/api/accounts/#account-type-schema
	SubType models.BankAccountSubType `json:"accountSubType" example:"checking" validate:"required"`
	// The original name of the bank account from when it was created. This name cannot be changed after the bank
	// account is created. This is primarily due to bank account's coming from a 3rd party provider like Plaid. But to
	// reduce the amount of logic in the application the same rule applies for manual links as well.
	PlaidName string `json:"originalName" example:"Checking Account #1"`
	// Official name is only used with bank accounts coming from Plaid. It is another name that Plaid uses for an
	// account.
	PlaidOfficialName string `json:"officialName" example:"US Bank - Checking Account"`
}

type BankAccountResponse struct {
	// Bank Account ID uniquely identifies this bank account within monetr.
	BankAccountId uint64 `json:"bankAccountId" example:"85473" validate:"required"`

	CreateBankAccountRequest

	// Last Updated is used to keep track of changes to this specific bank account. It is updated when the balances that
	// are provided by an external party (like Plaid) are changed. It is not necessarily changed when the Safe-To-Spend
	// balance changes or when other fields on the bank account change.
	LastUpdated time.Time `json:"lastUpdated" example:"2021-04-15T00:00:00-05:00"`
}

File: interface/src/models/BankAccount.ts
import { ID, idPrefix } from '@monetr/interface/models/ID';
import type Link from '@monetr/interface/models/Link';
import LunchFlowBankAccount from '@monetr/interface/models/LunchFlowBankAccount';
import PlaidBankAccount from '@monetr/interface/models/PlaidBankAccount';
import type { WithJsonValues } from '@monetr/interface/util/json';
import parseDate from '@monetr/interface/util/parseDate';

export type BankAccountStatus = 'unknown' | 'active' | 'inactive';

export enum BankAccountType {
  Depository = 'depository',
  Credit = 'credit',
  Loan = 'loan',
  Investment = 'investment',
  Other = 'other',
}

export enum BankAccountSubType {
  Checking = 'checking',
  Savings = 'savings',
  HSA = 'hsa',
  CD = 'cd',
  MoneyMarket = 'money market',
  PayPal = 'paypal',
  Prepaid = 'prepaid',
  CashManagement = 'cash management',
  EBT = 'ebt',
  CreditCard = 'credit card',
  Auto = 'auto',
  Other = 'other',
}

export default class BankAccount {
  readonly [idPrefix] = 'bac';

  readonly bankAccountId: ID<BankAccount>;
  readonly linkId: ID<Link>;
  readonly lunchFlowBankAccountId: ID<LunchFlowBankAccount> | null;
  mask: string | null;
  name: string;
  readonly originalName: string;
  readonly status: BankAccountStatus;
  accountType: BankAccountType;
  accountSubType: BankAccountSubType;
  currency: string;
  // Don't use these fields directly except when creating!
  currentBalance: number;
  availableBalance: number;
  limitBalance: number | null;
  readonly lastUpdated: Date;
  readonly createdAt: Date;
  readonly createdBy: string;
  readonly deletedAt: Date | null;

  readonly plaidBankAccount: PlaidBankAccount | null;
  readonly lunchFlowBankAccount: LunchFlowBankAccount | null;

  constructor(data: WithJsonValues<BankAccount>) {
    this.bankAccountId = ID.from(data.bankAccountId);
    this.linkId = ID.from(data.linkId);
    this.lunchFlowBankAccountId = data.lunchFlowBankAccountId ? ID.from(data.lunchFlowBankAccountId) : null;
    this.mask = data.mask ?? null;
    this.name = data.name;
    this.originalName = data.originalName;
    this.status = data.status;
    this.accountType = data.accountType;
    this.accountSubType = data.accountSubType;
    this.currency = data.currency;
    this.currentBalance = data.currentBalance;
    this.availableBalance = data.availableBalance;
    this.limitBalance = data.limitBalance ?? null;
    this.lastUpdated = parseDate(data.lastUpdated);
    this.createdAt = parseDate(data.createdAt);
    this.createdBy = data.createdBy;
    this.deletedAt = parseDate(data.deletedAt);
    this.plaidBankAccount = data.plaidBankAccount ? new PlaidBankAccount(data.plaidBankAccount) : null;
    this.lunchFlowBankAccount = data.lunchFlowBankAccount ? new LunchFlowBankAccount(data.lunchFlowBankAccount) : null;
  }
}
